pkg/snapshot/restorer/restorer.go

```diff
@@ -317,24 +326,260 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
 	return e, nil
 }
 
-// applyDeltaSnapshot applies the events from time sorted list of delta snapshot to etcd sequentially
-func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, snapList snapstore.SnapList) error {
+// applyDeltaSnapshots fetches the events from delta snapshots in parallel and applies them to the embedded etcd sequentially
+func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, ro RestoreOptions) error {
+	snapList := ro.DeltaSnapList
+	numMaxFetchers := ro.MaxFetchers
+
+	// no delta snapshots available
+	if len(snapList) == 0 {
+		return nil
+	}
+
 	firstDeltaSnap := snapList[0]
+
+	r.logger.Infof("Applying delta snapshot %s", path.Join(firstDeltaSnap.SnapDir, firstDeltaSnap.SnapName))
 	if err := r.applyFirstDeltaSnapshot(client, *firstDeltaSnap); err != nil {
 		return err
 	}
 	if err := verifySnapshotRevision(client, snapList[0]); err != nil {
 		return err
 	}
-	for _, snap := range snapList[1:] {
-		if err := r.applyDeltaSnapshot(client, *snap); err != nil {
-			return err
+
+	// no more delta snapshots available
+	if len(snapList) == 1 {
+		return nil
+	}
+
+	var (
+		remainingSnaps         = snapList[1:]
+		numSnaps               = len(remainingSnaps)
+		snapLocations          = make([]string, numSnaps)
+		errCh                  = make(chan error, 1)
+		snapInfoCh             = make(chan snapInfo, numSnaps)
+		fetchedEventsInfoCh    = make(chan eventsInfo, numSnaps)
+		applyInfoCh            = make(chan applyInfo, numSnaps)
+		numFetchers            = int(math.Min(float64(numMaxFetchers), float64(numSnaps)))
+		stopCh                 = make(chan bool, numFetchers+2) // used for stopping all routines (fetchers, persister and applier)
+		fetchersStoppedCh      = make(chan bool, numFetchers)   // used to indicate that fetchers have stopped
+		persistDeltasStoppedCh = make(chan bool, numFetchers)   // used to indicate that persister has stopped
+		applyStoppedCh         = make(chan bool, numFetchers)   // used to indicate that applier has stopped
+	)
+
+	go r.persistDeltaSnapshots(numSnaps, snapLocations, fetchedEventsInfoCh, applyInfoCh, errCh, stopCh, persistDeltasStoppedCh)
```

Comment:
Number of persister should be equal to number of fetchers, shouldn't it? Cause persister is also, doing diskIO task of disk which can be parallelised.

Reply:
During restoration, fetching and applying snaps usually take more time than persisting them. Applier cannot be parallelised, and fetcher is now parallelised. I didn't see the need to parallelise persister as well, as it wouldn't lead to any real performance improvement during restoration, but would still add the overhead of more go routines and synchronisation. What do you think?

Reply:
Fetcher and persister functionalities have now been combined to help curb memory over-utilisation, and also keeps clean, maintainable code.